### PR TITLE
Bugfix/additional delta operation for newline character

### PR DESCRIPTION
**Pull request by fschindler**
Fixes https://github.com/visualjerk/quill-magic-url/issues/36

### CODE REVIEW

File: src/index.test.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+          {
+            insert: 'www.hello@example.com',
+            attributes: { link: 'mailto:www.hello@example.com' },
+          },
+          { insert: ' !' },
+        ],
+      })
+    })
+  })
+})
```
Comment (visualjerk) [inline src/index.test.js:162]: ❤️  Thanks for the first test suite. Awesome work!

**Review by visualjerk — CHANGES_REQUESTED** [review]
Thanks for the bugfix and BIIIIG thanks for adding unit tests. I was too lazy to do that in the past.

I really like how you handle the different patterns by tracing the regex index.

If you don't mind, it would be nice to simplify the while loop a bit (see comments). That would make it easier to reason about for future developers.

File: src/index.js [hunk 2]
```diff
@@ -27,25 +27,44 @@ export default class MagicUrl {
...
+          }
         }
-        if (urlMatches && urlMatches.length) {
-          urlMatches.forEach(match => addMatchToDelta(match, this.urlNormalizer))
-        } else if (mailMatches && mailMatches.length) {
-          mailMatches.forEach(match => addMatchToDelta(match, this.mailNormalizer))
+        if (urlResult === null && mailResult === null) {
+          const tail = node.data.substring(index)
+          newDelta.insert(tail);
         }
```
[the remark this answers is not in the document]
Comment (fschindler) [reply src/index.js]: You're right, good catch! I've moved this code out of the loop.

File: src/index.js [hunk 3]
```diff
@@ -27,25 +27,44 @@ export default class MagicUrl {
...
+        return regExp.exec(node.data)
+      }
+      while(urlResult !== null || mailResult !== null) {
+        if (urlResult === null) {
+          mailResult = handleMatch(mailResult, mailRegExp, this.mailNormalizer)
+        } else if (mailResult === null) {
+          urlResult = handleMatch(urlResult, urlRegExp, this.urlNormalizer)
+        } else if (urlRegExp.lastIndex <= mailRegExp.lastIndex) {
+          urlResult = handleMatch(urlResult, urlRegExp, this.urlNormalizer)
+          while (mailResult !== null && mailRegExp.lastIndex < urlRegExp.lastIndex) {
```
[the remark this answers is not in the document]
Comment (fschindler) [reply src/index.js]: I've improved the tests and found/fixed some bugs. But we still need this.

Whenever the url and mail pattern matched a the same substring we only want to handle one of these matches. The other RegExp should be forwarded to the next match (after this substring). E.g. handle `urlResult` and forward `mailRegExp` until `mailResult.index >= urlRegExp.lastIndex`.

Hopefully https://github.com/visualjerk/quill-magic-url/pull/37/files#diff-38a2b85ac1b61b44eb1945e128241cb5ef01e5646cf06b586f5c0114b153fa0eR139 and https://github.com/visualjerk/quill-magic-url/pull/37/files#diff-38a2b85ac1b61b44eb1945e128241cb5ef01e5646cf06b586f5c0114b153fa0eR139 give a short demonstration of this.

**Comment by fschindler** [comment]
Hi @visualjerk, thanks for your review and valuable feedback!

I've added two more commits.

File: src/index.test.js [hunk 4]
```diff
@@ -136,23 +136,40 @@ describe('MagicUrl', () => {
       })
     })
 
-    test('should parse same url multiple times', () => {
+    test('should parse url instead of mail', () => {
+      // the url is part of an mail and both start at the same index
```
Comment (visualjerk) [inline src/index.test.js:140]: Haven't even thought about those edge cases 😅 

Actually in that case I would expect to only get the mail and not the url, because as you correctly state "the url is part of an mail".

**Review by visualjerk — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (visualjerk) [reply src/index.js]: Thanks for clarification 👍

File: src/index.test.js (hunk 4 above)
Comment (fschindler) [reply src/index.test.js]: I agree on the expected behaviour. So I've switch the priority of the url and mail matching.

My initial idea was to make the changes as small as possible. And the current behaviour is more like favouring url against mail.

Comment (visualjerk) [reply src/index.test.js]: Nice, thanks for the changes!

**Review by visualjerk — APPROVED** [review]
Thanks for the great MR and especially for adding unit tests.

The bugfix will be released in version 3.0.1.
